### PR TITLE
Fix scrollable `TabBar` jittering

### LINKED ISSUE #150000: TabBar with isScrollable set to true is broken
### Steps to reproduce

1. Run example code using Flutter 3.23.0-0.1.pre on Android Emulator - Pixel 7 API 34 or real Android device
2. Swipe the TabBarView to go to Tab 3 or greater.

### Expected results

The TabBar should act normally.

### Actual results

The TabBar animates to start position and back. See attached video for demostration.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage>
    with SingleTickerProviderStateMixin {
  late TabController _tabController;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(length: 8, vsync: this);
  }

  @override
  void dispose() {
    _tabController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        bottom: TabBar(
          controller: _tabController,
          dragStartBehavior: DragStartBehavior.down,
          isScrollable: true,
          tabs: [
            for (int i = 0; i < 8; i++) Tab(text: 'Tab $i'),
          ],
        ),
      ),
      body: TabBarView(
        controller: _tabController,
        children: [
          for (int i = 0; i < 8; i++)
            Center(
              child: Text('Content of Tab $i'),
            ),
        ],
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/8590720/ec3a127f-f877-4e9a-aaa2-0f5adffb9e06

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel beta, 3.23.0-0.1.pre, on macOS 14.5 23F79 darwin-arm64, locale en-GB)
    • Flutter version 3.23.0-0.1.pre on channel beta at /Users/per/dev/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2feea7a407 (5 days ago), 2024-06-06 10:19:10 +0700
    • Engine revision bb10c54666
    • Dart version 3.5.0 (build 3.5.0-180.3.beta)
    • DevTools version 2.36.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/per/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.90.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (4 available)
    • sdk gphone64 arm64 (mobile)     • emulator-5554         • android-arm64  • Android 14 (API 34) (emulator)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.5 23F79 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.5 23F79 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 125.0.6422.142
    ! Error: Browsing on the local area network for Per’s iPhone. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes [TabBar with isScrollable set to true is broken](https://github.com/flutter/flutter/issues/150000)

This regressed due to a tiny mistake in https://github.com/flutter/flutter/pull/146293

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'dart:ui';

import 'package:flutter/material.dart';

/// Flutter code sample for [TabBar].

void main() => runApp(const TabBarApp());

class TabBarApp extends StatelessWidget {
  const TabBarApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: TabBarExample(),
    );
  }
}

class TabBarExample extends StatelessWidget {
  const TabBarExample({super.key});

  @override
  Widget build(BuildContext context) {
    final List<Tab> tabs =
        List<Tab>.generate(20, (int index) => Tab(text: 'Tab $index'));

    return ScrollConfiguration(
      behavior: ScrollConfiguration.of(context)
          .copyWith(dragDevices: <PointerDeviceKind>{
        PointerDeviceKind.touch,
        PointerDeviceKind.mouse,
      }),
      child: DefaultTabController(
        length: tabs.length,
        child: Scaffold(
          appBar: AppBar(
            title: const Text('TabBar Sample'),
            bottom: TabBar(
              isScrollable: true,
              tabs: tabs,
              tabAlignment: TabAlignment.start,
            ),
          ),
          body: TabBarView(
            children: <Widget>[
              for (int i = 0; i < tabs.length; i++)
                Center(
                  child: Text('Page $i'),
                ),
            ],
          ),
        ),
      ),
    );
  }
}

```

</details>

### Before

https://github.com/flutter/flutter/assets/48603081/b7aa98a2-a6a5-431e-8327-859a11efa129

### After

https://github.com/flutter/flutter/assets/48603081/0435719f-03d4-4d76-8b5a-532894fcf4a3

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by PerLycke** [comment]
Even though this fix works to some extent, it doesn't fix "TabBar with isScrollable set to true is broken" as with the changes proposed in this PR the TabBar jumps after swiping to Tab 3 and back to Tab 2:

https://github.com/flutter/flutter/assets/8590720/214a8321-1b83-4a9d-b9d4-bf1e83e01dc5

**Comment by TahaTesser** [comment]
@PerLycke 
Thank you so much for testing the PR code. I'm out of office today, I'll address your issue first thing tomorrow.

**Comment by PerLycke** [comment]
@TahaTesser I can confirm it now works as expected 👍

**Review by justinmc — APPROVED** [review]
Tricky, good catch. LGTM :+1: 

FYI @nate-thegrate

File: packages/flutter/lib/src/material/tabs.dart [hunk 1]
```diff
@@ -1548,13 +1548,12 @@ class _TabBarState extends State<TabBar> {
 
     final double index = _controller!.index.toDouble();
     final double value = _controller!.animation!.value;
-    final double offset = switch (value - index) {
-      -1.0 || 1.0 => leadingPosition ?? middlePosition,
-      0 => middlePosition,
-      < 0 when leadingPosition == null => middlePosition,
-      > 0 when trailingPosition == null => middlePosition,
-      < 0 => lerpDouble(middlePosition, leadingPosition, index - value)!,
-      _   => lerpDouble(middlePosition, trailingPosition, value - index)!,
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/tabs.dart:1557]: Thanks for catching my mistake here!

(And thanks @justinmc for letting me know!)

File: packages/flutter/lib/src/material/tabs.dart [hunk 2]
```diff
@@ -1548,13 +1548,12 @@ class _TabBarState extends State<TabBar> {
...
+    final double offset = switch ((value, index)) {
+      (_, _) when value == index - 1.0 =>  leadingPosition ?? middlePosition,
+      (_, _) when value == index + 1.0 =>  trailingPosition ?? middlePosition,
+      (_, _) when value == index =>  middlePosition,
+      (_, _) when value < index =>  leadingPosition == null ? middlePosition : lerpDouble(middlePosition, leadingPosition, index - value)!,
+      (_, _) =>  trailingPosition == null ? middlePosition : lerpDouble(middlePosition, trailingPosition, value - index)!,
     };
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/tabs.dart:1557]: It feels a bit weird to pass a record into a switch expression and then not use it.

```suggestion
    final double offset = switch (value - index) {
      -1.0 => leadingPosition ?? middlePosition,
      1.0  => trailingPosition ?? middlePosition,
      0    => middlePosition,
      < 0  => leadingPosition == null ? middlePosition : lerpDouble(middlePosition, leadingPosition, index - value)!,
      _    => trailingPosition == null ? middlePosition : lerpDouble(middlePosition, trailingPosition, value - index)!,
    };
```

I believe the logic is identical, based on a local test run.

image

**Review by nate-thegrate — APPROVED** [review]
I really appreciate the effort put into correcting a previous error I made.

If we want to just merge right away, that's good with me 👍

File: packages/flutter/lib/src/material/tabs.dart (hunk 2 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/tabs.dart]: This is better, thanks!

**Pull request merged**